linera-chain/src/certificate.rs

```diff
@@ -0,0 +1,307 @@
+            round: self.round,
+            signatures: self.signatures.clone(),
+        }
+    }
+}
+
+impl<T: Debug> Debug for CertificateT<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CertificateT")
+            .field("value", &self.value)
+            .field("round", &self.round)
+            .field("signatures", &self.signatures)
+            .finish()
+    }
+}
+
+// NOTE: For backwards compatiblity reasons we serialize the new `Certificate` type as the old
+// one. Otherwise we would be breaking the RPC API schemas. We can't implement generic serialization
+// for `Certificate<T>` since only specific `T`s have corresponding `CertificateValue` variants.
+impl Serialize for ValidatedBlockCertificate {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let cert = Certificate::from(self.clone());
+        cert.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for ValidatedBlockCertificate {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Certificate::deserialize(deserializer).map(ValidatedBlockCertificate::from)
+    }
+}
+
+#[cfg(with_testing)]
+impl<T: Eq + PartialEq> Eq for CertificateT<T> {}
+#[cfg(with_testing)]
+impl<T: Eq + PartialEq> PartialEq for CertificateT<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.value.hash == other.value.hash
+            && self.round == other.round
+            && self.signatures == other.signatures
+    }
+}
+
+impl From<ConfirmedBlockCertificate> for Certificate {
+    fn from(cert: ConfirmedBlockCertificate) -> Certificate {
+        let ConfirmedBlockCertificate {
+            value,
+            round,
+            signatures,
+        } = cert;
+        Certificate::new(
+            HashedCertificateValue::new_confirmed(value.into_inner().into_inner()),
+            round,
+            signatures,
+        )
+    }
+}
+
+impl From<ValidatedBlockCertificate> for Certificate {
+    fn from(cert: ValidatedBlockCertificate) -> Certificate {
+        let ValidatedBlockCertificate {
+            value,
+            round,
+            signatures,
+        } = cert;
+        Certificate::new(
+            HashedCertificateValue::new_validated(value.into_inner().into_inner()),
+            round,
+            signatures,
+        )
+    }
+}
+
+// In practice, it should be HashedCertificateValue = Hashed<CertificateValue>
+// but [`HashedCertificateValue`] is used in too many places to change it now.
+pub struct Hashed<T> {
+    value: T,
+    hash: CryptoHash,
+}
+
+impl<T> Hashed<T> {
+    // Note on usage: This method is unsafe because it allows the caller to create a Hashed
+    // with a hash that doesn't match the value. This is necessary for the rewrite state when
+    // signers sign over old `Certificate` type.
+    pub fn unsafe_new(value: T, hash: CryptoHash) -> Self {
```

Comment:
Maybe "unchecked" or "unverified" instead of "unsafe", to distinguish from `unsafe` Rust code?

Reply:
Changing to unchecked_new

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/2803/commits/718f2bf2d0a65301bf6b5ca8eef24099a8a3235a